### PR TITLE
Knn clustering

### PR DESCRIPTION
**Pull request by oeldaghar**

### CODE REVIEW

File: src/spac/transformations.py [hunk 1]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+    `.uns["knn_features"]`
+        The features used to calculate the knn clusters
+
+    Parameters
+    ----------
+    adata : anndata.AnnData
+       The AnnData object.
+
+    features : list of str
+        The variables that would be included in creating the phenograph
```
Comment (georgezakinih) [inline src/spac/transformations.py:132]: @oeldaghar, the cluster is not phenograph

File: src/spac/transformations.py [hunk 2]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+        clusters.
+    
+    annotation : str
+        The name of the annotation used for classifying the data
+
+    layer : str, optional
+        The layer to be used in calculating the phengraph clusters.
```
Comment (georgezakinih) [inline src/spac/transformations.py:139]: @oeldaghar, same comment about phenograph

File: src/spac/transformations.py [hunk 3]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+
+    k : int, optional
```
Comment (georgezakinih) [inline src/spac/transformations.py:141]: @oeldaghar, is it creating a graph? I believe it is a ball tree.

File: src/spac/transformations.py [hunk 4]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+    _validate_transformation_inputs(
+        adata=adata,
+        layer=layer,
+        associated_table=associated_table,
+        features=features,
+        annotation=annotation,
+    )
+
+    if not isinstance(k, int) or k <= 0:
+        raise ValueError("`k` must be a positive integer")
```
Comment (georgezakinih) [inline src/spac/transformations.py:170]: @oeldaghar , please add in the error message the received value as it is a hint for the user.

File: src/spac/transformations.py [hunk 5]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+
+    # boolean mask for labeled data
+    annotation_data = adata.obs[annotation]
+    annotation_mask = annotation_data != missing_label
+    # handle np.nan values
+    annotation_mask &= pd.notnull(annotation_data)
+
+    # check that annotation is non-trivial
+    if all(annotation_mask):
+        raise ValueError("All cells are labeled. Please provide a mix of labeled and unlabeled data.")
```
Comment (georgezakinih) [inline src/spac/transformations.py:187]: @oeldaghar, please mention the "annotation" that has been used. So something like: f'All cells are labeled in the annotation "{annotation". ..."

File: src/spac/transformations.py [hunk 6]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+    elif not any(annotation_mask):
+        raise ValueError("No cells are labeled. Please provide a mix of labeled and unlabeled data.")
```
Comment (georgezakinih) [inline src/spac/transformations.py:189]: @oeldaghar, same as the above comment.

File: src/spac/transformations.py [hunk 7]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+
+    # fit knn classifier to labeled data and predict on full dataset     
+    data_labeled = data[annotation_mask]
+    annotation_labeled = np.array(annotation_data[annotation_mask], dtype=int)
```
Comment (georgezakinih) [inline src/spac/transformations.py:193]: What is the values in adata.obs[annotation] is a string?

File: src/spac/transformations.py [hunk 8]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+    
+    classifier = KNeighborsClassifier(n_neighbors = k, **kwargs)
+    classifier.fit(data_labeled, annotation_labeled)
+    knn_predict = classifier.predict(data)
```
Comment (georgezakinih) [inline src/spac/transformations.py:197]: @oeldaghar, why predict on the data that was already labeled?

File: src/spac/transformations.py [hunk 9]
```diff
@@ -102,6 +103,103 @@ def phenograph_clustering(
...
+    # store output 
+    adata.obs[output_annotation] =  pd.Categorical(knn_predict)
+    adata.uns["knn_features"] = features
```
Comment (georgezakinih) [inline src/spac/transformations.py:200]: @oeldaghar, suggestion, instead of adata.uns["knn_features"], use adata.uns[output_annotation]

File: tests/test_transformations/test_knn_clustering.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
+import unittest
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+from sklearn.datasets import load_iris
+from spac.transformations import knn_clustering
+
+
+class TestKnnClustering(unittest.TestCase):
+    def setUp(self):
```
Comment (georgezakinih) [inline tests/test_transformations/test_knn_clustering.py:10]: @oeldaghar, can we start with a very simple case of minimum number of points (e.g., 4 points, two labeled and two unlabled, with K=1), then elaborate that use case to cover more corner cases?

File: tests/test_transformations/test_knn_clustering.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
+import unittest
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+from sklearn.datasets import load_iris
```
Comment (georgezakinih) [inline tests/test_transformations/test_knn_clustering.py:5]: @oeldaghar, do we need that dataset?

File: tests/test_transformations/test_knn_clustering.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        knn_clustering(
+            adata=self.syn_data,
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer=None,
+        )
+        self.assertIn("knn", self.syn_data.obs)
+        self.assertEqual(self.syn_data.uns["knn_features"], self.syn_features)
+
+    def test_invalid_k(self):
```
Comment (georgezakinih) [inline tests/test_transformations/test_knn_clustering.py:115]: @oeldaghar, please also check that the appropriate error message is generated.

File: tests/test_transformations/test_knn_clustering.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        knn_clustering(
+            adata=self.syn_data,
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer=self.layer,
+            output_annotation=output_annotation_name,
+        )
+        self.assertIn(output_annotation_name, self.syn_data.obs)
+
+    def test_layer_none_case(self):
```
Comment (georgezakinih) [inline tests/test_transformations/test_knn_clustering.py:104]: @oeldaghar, I suggest to put that test as part of _select_input_features if that is not already there.

File: tests/test_transformations/test_knn_clustering.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+        with self.assertRaises(ValueError):
+            knn_clustering(
+                adata=self.syn_data,
+                features=self.syn_features,
+                annotation=self.annotation,
+                layer=self.layer,
+                k="invalid",
+            )
+
+    def test_trivial_label(self):
```
Comment (georgezakinih) [inline tests/test_transformations/test_knn_clustering.py:127]: @oeldaghar, please check the error message too.

File: tests/test_transformations/test_knn_clustering.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer="counts",
+            k=50,
+        )
+
+        self.assertIn("knn", self.syn_data.obs)
+        self.assertEqual(len(np.unique(self.syn_data.obs["knn"])), 2)
+
+    def test_associated_features(self):
```
Comment (georgezakinih) [inline tests/test_transformations/test_knn_clustering.py:160]: @oeldaghar, I believe that test is already covered part of _select_input_features.

**Review by georgezakinih — CHANGES_REQUESTED** [review]
@oeldaghar, Thanks for putting the PR. I shared with you few comments. It's a good first pass for adding functionality to SPAC. We can also schedule a code review session if you want to discuss these comments.

**Comment by oeldaghar** [comment]
@akolliegbo could you take a look at this and make the requested changes. we can schedule time to meet if it helps digests the request
